File: content/tracks/javascript/concepts/02-functions/functions_test.js
const assert = require("assert");
const { multiply, sum, compose } = require("./stub");

let passed = 0, failed = 0;
function test(name, fn) {
  try { fn(); console.log(`  PASS: ${name}`); passed++; }
  catch (e) { console.log(`  FAIL: ${name} — ${e.message}`); failed++; }
}

test("multiply(3, 4)",      () => assert.strictEqual(multiply(3, 4),  12));
test("multiply(5)",         () => assert.strictEqual(multiply(5),      5));
test("multiply(0, 100)",    () => assert.strictEqual(multiply(0, 100), 0));
test("multiply(-2, 3)",     () => assert.strictEqual(multiply(-2, 3), -6));

test("sum()",               () => assert.strictEqual(sum(),          0));
test("sum(1)",              () => assert.strictEqual(sum(1),         1));
test("sum(1,2,3)",          () => assert.strictEqual(sum(1, 2, 3),   6));
test("sum(1,2,3,4,5)",      () => assert.strictEqual(sum(1,2,3,4,5), 15));
test("sum negative",        () => assert.strictEqual(sum(-1,-2,-3),  -6));

const double = (x) => x * 2;
const inc    = (x) => x + 1;
test("compose(double, inc)(3)", () => assert.strictEqual(compose(double, inc)(3), 8));
test("compose(inc, double)(3)", () => assert.strictEqual(compose(inc, double)(3), 7));
test("compose: identity",       () => assert.strictEqual(compose((x)=>x, (x)=>x)(42), 42));

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/python/concepts/02-functions/functions_test.py
import sys
from stub import total, make_multiplier, pipeline

passed = 0
failed = 0


def assert_eq(actual, expected):
    assert actual == expected, f"expected {expected!r}, got {actual!r}"


def assert_close(actual, expected, tol=1e-9):
    assert abs(actual - expected) < tol, f"expected ~{expected}, got {actual}"


def test(name: str, fn):
    global passed, failed
    try:
        fn()
        print(f"  PASS: {name}")
        passed += 1
    except Exception as e:
        print(f"  FAIL: {name} — {e}")
        failed += 1


# total
test("total: multiple args",  lambda: assert_close(total(1, 2, 3), 6.0))
test("total: single arg",     lambda: assert_close(total(5), 5.0))
test("total: no args",        lambda: assert_close(total(), 0.0))
test("total: floats",         lambda: assert_close(total(1.5, 2.5), 4.0))
test("total: negative",       lambda: assert_close(total(-1, -2, 3), 0.0))

# make_multiplier
test("make_multiplier: double",  lambda: assert_close(make_multiplier(2)(5), 10.0))
test("make_multiplier: triple",  lambda: assert_close(make_multiplier(3)(4), 12.0))
test("make_multiplier: zero",    lambda: assert_close(make_multiplier(0)(99), 0.0))
test("make_multiplier: float",   lambda: assert_close(make_multiplier(0.5)(10), 5.0))
test("make_multiplier: returns callable", lambda: assert_eq(callable(make_multiplier(2)), True))

# pipeline
double = make_multiplier(2)
triple = make_multiplier(3)
test("pipeline: no fns",           lambda: assert_close(pipeline(7), 7.0))
test("pipeline: single fn",        lambda: assert_close(pipeline(3, double), 6.0))
test("pipeline: two fns",          lambda: assert_close(pipeline(3, double, triple), 18.0))
test("pipeline: order matters",    lambda: assert_close(pipeline(2, make_multiplier(3), make_multiplier(4)), 24.0))
test("pipeline: add then multiply", lambda: assert_close(
    pipeline(5, lambda x: x + 1, make_multiplier(2)), 12.0
))

print(f"\n{passed} passed, {failed} failed")
if failed:
    sys.exit(1)
